packages/muelu/src/Graph/UncoupledAggregation/MueLu_AggregationPhase2aAlgorithm_kokkos_def.hpp

```diff
@@ -63,78 +63,108 @@
 
-    LO numLocalNodes      = procWinner.size();
+    LO numLocalNodes      = numRows;
     LO numLocalAggregated = numLocalNodes - numNonAggregatedNodes;
 
     const double aggFactor = 0.5;
     double       factor    = as<double>(numLocalAggregated)/(numLocalNodes+1);
     factor = pow(factor, aggFactor);
 
-    int              aggIndex = -1;
-    size_t           aggSize  =  0;
-    std::vector<int> aggList(graph.getNodeMaxNumRowEntries());
-
-    for (LO rootCandidate = 0; rootCandidate < numRows; rootCandidate++) {
-      if (aggStat[rootCandidate] != READY)
-        continue;
-
-      aggSize = 0;
-
-      auto neighOfINode = graph.getNeighborVertices(rootCandidate);
-
-      LO numNeighbors = 0;
-      for (int j = 0; j < as<int>(neighOfINode.length); j++) {
-        LO neigh = neighOfINode(j);
-
-        if (neigh != rootCandidate) {
-          if (graph.isLocalNeighborVertex(neigh) && aggStat[neigh] == READY) {
-            // If aggregate size does not exceed max size, add node to the tentative aggregate
-            // NOTE: We do not exit the loop over all neighbours since we have still
-            //       to count all aggregated neighbour nodes for the aggregation criteria
-            // NOTE: We check here for the maximum aggregation size. If we would do it below
-            //       with all the other check too big aggregates would not be accepted at all.
-            if (aggSize < as<size_t>(maxNodesPerAggregate))
-              aggList[aggSize++] = neigh;
-          }
-
-          numNeighbors++;
-        }
-      }
-
-      // NOTE: ML uses a hardcoded value 3 instead of MinNodesPerAggregate
-      if (aggSize > as<size_t>(minNodesPerAggregate) &&
-          aggSize > factor*numNeighbors) {
-        // Accept new aggregate
-        // rootCandidate becomes the root of the newly formed aggregate
-        aggregates.SetIsRoot(rootCandidate);
-        aggIndex = numLocalAggregates++;
-
-        for (size_t k = 0; k < aggSize; k++) {
-          aggStat     [aggList[k]] = AGGREGATED;
-          vertex2AggId[aggList[k]] = aggIndex;
-          procWinner  [aggList[k]] = myRank;
-        }
-
-        numNonAggregatedNodes -= aggSize;
-      }
+    Kokkos::View<LO, memory_space> numLocalAggregates("numLocalAggregates");
+    typename Kokkos::View<LO, memory_space>::HostMirror h_numLocalAggregates =
+      Kokkos::create_mirror_view(numLocalAggregates);
+    h_numLocalAggregates() = aggregates.GetNumAggregates();
+    Kokkos::deep_copy(numLocalAggregates, h_numLocalAggregates);
+
+    // Now we create new aggregates using root nodes in all colors other than the first color,
+    // as the first color was already exhausted in Phase 1.
+    Kokkos::TeamPolicy<execution_space> outerPolicy(numRows, Kokkos::AUTO);
+    typedef typename Kokkos::TeamPolicy<execution_space>::member_type  member_type;
+    for(int color = 1; color < numColors + 1; ++color) {
```

Comment:
@brian-kelley  yes this should be a two there, I had in mind that colors are starting at 0 but that's actually 1, I'll make that modification now.
I will work a bit on phase 3 today since it is way to slow on GPU which is annoying...

Reply:
@brian-kelley  actually I remembered that you are working on phase 3 already so I will leave as is for now until we decide to merge things together.